lisa/sut_orchestrator/baremetal/cluster/idrac.py

```diff
@@ -214,30 +240,139 @@ def _wait_for_completion(self, response: Any, timeout: int = 600) -> None:
+        # Poll manager until Enabled (up to 2 minutes)
+        start_time = time.time()
+        while time.time() - start_time < IDRAC_RESET_TIMEOUT:
+            try:
+                mgr_state = self.redfish_instance.get(
+                    "/redfish/v1/Managers/iDRAC.Embedded.1"
+                ).dict
+                if mgr_state.get("Status", {}).get("State") == "Enabled":
+                    self._log.info("iDRAC reset completed successfully")
+                    return
+            except Exception:
+                # iDRAC may be restarting, ignore connection errors
+                pass
+            time.sleep(2)
+
+        raise LisaException("iDRAC did not come back after Manager.Reset")
+
+    def _insert_virtual_media(self, iso_http_url: str) -> None:
+        """Insert virtual media with robust error handling and retry logic."""
+        self._log.info(f"Inserting virtual media from URL: {iso_http_url}")
+
+        # Step 1: Ensure all virtual media is ejected
+        self._ensure_vm_cleared()
+
+        # Step 2: Sleep to let iDRAC drop any stale NFS/HTTP handles
+        time.sleep(5)
+
+        # Step 3: Try insert with retries and iDRAC reset if needed
+        max_attempts = 3
+        for attempt in range(1, max_attempts + 1):
+            try:
+                # Explicitly specify NFS protocol (helps across firmware revisions)
+                body = {"Image": iso_http_url, "TransferProtocolType": "NFS"}
+                response = self.redfish_instance.post(
+                    "/redfish/v1/Managers/iDRAC.Embedded.1/VirtualMedia/CD/"
+                    "Actions/VirtualMedia.InsertMedia",
+                    body=body,
+                )
+                self._wait_for_completion(response)
+
+                # Step 4: Verify insertion by polling CD state
+                start_time = time.time()
+                while time.time() - start_time < VIRTUAL_MEDIA_INSERTION_POLL_TIMEOUT:
+                    if self._get_vm_state("CD").get("Inserted"):
+                        self._log.info("Virtual media insertion completed successfully")
+                        return
+                    time.sleep(1)
+
+                raise LisaException(
+                    "Insert reported success but CD not showing as Inserted"
+                )
+
+            except LisaException as e:
+                error_msg = str(e)
+                # Check for RAC0904 or reachability errors
+                is_reachability_error = (
+                    "RAC0904" in error_msg or "not accessible or reachable" in error_msg
+                )
+
+                if is_reachability_error and attempt < max_attempts:
+                    self._log.debug(
+                        f"RAC0904/reachability error on attempt {attempt}. "
+                        "Ejecting media, resetting iDRAC, and retrying..."
+                    )
+                    # Clear virtual media again
+                    self._ensure_vm_cleared()
```

Comment:
Why does it need to clean up and insert again? If the first attempt fails, why would subsequent attempts succeed? Add comments above to explain this.

Reply:
updated comment ,pls check